tapfreighter/wallet.go

```diff
@@ -1231,28 +1211,139 @@ func removeActiveCommitments(inputCommitment *commitment.TapCommitment,
-		key := vIn.Asset().TapCommitmentKey()
-		assetCommitment, ok := passiveCommitments[key]
-		if !ok {
+	for _, vPkt := range vPackets {
+		for _, vIn := range vPkt.Inputs {
+			key := vIn.Asset().TapCommitmentKey()
+			assetCommitment, ok := passiveCommitments[key]
+			if !ok {
+				continue
+			}
+
+			err := removeAsset(assetCommitment, vIn.Asset(), key)
+			if err != nil {
+				return nil, fmt.Errorf("unable to "+
+					"delete asset: %w", err)
+			}
+		}
+	}
+
+	return passiveCommitments, nil
+}
+
+// determinePassiveAssetAnchorOutput determines the best anchor output to attach
+// passive assets to. If no suitable output is found, a new anchor output is
+// created.
+func (f *AssetWallet) determinePassiveAssetAnchorOutput(ctx context.Context,
+	activePackets []*tappsbt.VPacket) (*keychain.KeyDescriptor, uint32,
+	error) {
+
+	var (
+		maxAnchorOutputIndex uint32
+		candidates           []*tappsbt.VOutput
+		candidateDescriptors []*keychain.KeyDescriptor
+	)
+	for idx := range activePackets {
+		vPkt := activePackets[idx]
+
+		for _, vOut := range vPkt.Outputs {
+			anchorKeyDesc, err := vOut.AnchorKeyToDesc()
+			if err != nil {
+				// We can't determine the key descriptor for
+				// this output, so we'll skip it as it very
+				// likely doesn't belong to us then.
+				continue
+			}
+
+			// Ignore any anchor outputs that are not local to us.
+			if !f.cfg.KeyRing.IsLocalKey(ctx, anchorKeyDesc) {
+				continue
+			}
+
+			candidates = append(candidates, vOut)
+			candidateDescriptors = append(
+				candidateDescriptors, &anchorKeyDesc,
+			)
+		}
+
+		// In case we need to create a new anchor output, we'll want to
+		// know the next anchor output index we can use. So we find the
+		// maximum currently used index.
+		for _, vOut := range vPkt.Outputs {
+			if vOut.AnchorOutputIndex > maxAnchorOutputIndex {
+				maxAnchorOutputIndex = vOut.AnchorOutputIndex
+			}
+		}
+	}
+
+	// From the candidates, we want to select one in descending order of
+	// preference:
+	//	1. A split root output (as that's usually the change).
+	//	2. A normal output to a new script key (probably a full-value
+	//	   send to ourselves).
+	//
+	// We start with the split root outputs:
+	for idx, vOut := range candidates {
+		if vOut.Type == tappsbt.TypeSplitRoot {
+			return candidateDescriptors[idx], vOut.AnchorOutputIndex,
+				nil
+		}
+	}
+
+	// We're still here, so let's try to find a normal output to a new
+	// script key.
+	for idx, vOut := range candidates {
+		// Skip any incomplete outputs that would cause the below
+		// statement to panic.
+		if vOut.Asset == nil || len(vOut.Asset.PrevWitnesses) == 0 ||
+			vOut.ScriptKey.PubKey == nil {
+
 			continue
 		}
 
-		err := removeAsset(assetCommitment, vIn.Asset(), key)
-		if err != nil {
-			return nil, fmt.Errorf("unable to "+
-				"delete asset: %w", err)
+		fromKey := vOut.Asset.PrevWitnesses[0].PrevID.ScriptKey
+		toKey := asset.ToSerialized(vOut.ScriptKey.PubKey)
+		if fromKey != toKey {
```

Comment:
So this is trying to rule out the case of anchoring to a self-send w/ addr re-use?

Reply:
Yes. Basically we're trying to find an "active" transfer that goes to an anchor internal key we own, so we can commit the passive assets there. And the only two heuristics for finding out if a transfer is "active" is that either it's a split or it goes to a new script key (full value send to yourself).
So worst case is we create a new anchor output even though we don't necessarily need to.

Reply:
Although, thinking back to why I created this logic: The goal was to ultimately remove the conceptual separation of active and passive assets in the RPCs and just identify them based on the heuristic described above. But since we currently still have the distinction between active/passive in the RPCs, maybe we don't need this more complicated logic and just pick any anchor output that goes back to our wallet (or create one if there is none).